### PR TITLE
Add storage RabbitMQ

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add storage RabbitMQ.

Detailed description / Documentation draft:

CRATE TABLE query parameters:

- rabbitmq_host_port
- rabbitmq_routing_key_list
- rabbitmq_exchange_name
- rabbitmq_exchange_type
- rabbitmq_format
- rabbitmq_row_delimiter
- rabbitmq_num_consumers
- rabbitmq_num_queues
- rabbitmq_transactional_channel

Some explanation:

The routing key of the message and exchange name are parameters that are set while publishing messages from any RabbitMQ client. Therefore, they are speicified by the client.

All exchange types, known in RabbitMQ, are supported: direct-exchange, fanout-exchange, topic-exchange, headers-exchange, consistent-hash-exchange. The preferred exchange type is specified by the client in exchange_type parameter.

There can be no more than one exchange per table. Also one exchange can be shared between multiple tables - it enables routing into multiple tables at the same time.

If exchange type is set to 'direct', then messages go to table(s) with a routing_key parameter that exactly matches the routing key of the message.

If exchange type is set to 'fanout', it routes received messages to all tables that are bound to it regardless of routing keys. (To table(s), where the exchange_name is the same.)

If exchange type is set to 'topic', then messages are routed to one or many tables based on a matching between a message routing key and a pattern, which is set in routing_key_list parameter. By RabbitMQ documentation, if messages are sent to topic-exchange, then a routing key must be a list of words, delimited by a dot. The routing patterns may contain “\*”. For example, if a pattern (specified in routing_key_list) of some tables is '\*.logs', then messages with a routing key with suffix '.logs' will be sent to all those tables. Patterns can be complicated (like 'agreements.\*.\*.data.'). Patterns can also contain '#' -  indicates a match of zero or more words. For more details see RabbitMQ documentation.

If exchange type is set to 'headers', the routing is even more flexible. If using headers-exchange, routing key can be a map (dictionary). For example, if (key = value) headers of published messages are 'format=logs', 'type=report', 'year=2020' with setting 'x-match=all', then these messages will be routed to all tables, where routing_key_list contains all those matches. If a setting is 'x-match=any', then those messages are send to all tables where at least one dictionary value matches its key. The setting is passed in the same list as headers:
rabbitmq_routing_key_list = 'x-match=all,format=logs,type=report,year=2020'.

If exchange type is set to 'consistent_hash', which is a special sharding exchange, - there will be an even distribution of messages between all tables, where the exchange_name is the same. Note that, by RabbitMQ documentation, if using consistent-hash-exchange, by default it will distribute messages based on a hash value of a routing key. Therefore, the key must be a string integer, randomised for every batch. 

Multiple routing keys for each table are supported. They are specified in a list of keys, separated by commas: rabbitmq_routing_key_list = 'key1,key2,key3,key4,key5'. In case of topic-exchange it can be a list of patterns, separated by commas: rabbitmq_routing_key_list = '\*.logs,data.\*,\*.work.data.*.logs'. In case of headers-exchange it can be a list of (key = value) headers: rabbitmq_routing_key_list = 'x-match=any,format=logs,type=report,year=2020'.

If parameter num_consumers is set, then in current implementation the exchange, specified by the client, is bound to a consistent-hash exchange (by the specified routing keys) and this exchange will make an even distribution between all (concurrent) consumers within one table, because, without this binding to consistent-hash exchange, it is impossible to ensure that one message is received no more than once, as it is the only way to enable sharding between queues (~ consumers)) with the same routing key. 

If parameter num_queues (number of queues per consumer) is set, then distribution of messages is done between all queues of all consumers. It is worth setting num_queues parameter as one queue can handle up to 50K messages and it is in general highly recommended to keep queues as short as possible to increase throughput. By default, there is one unique queue for each consumer. Also there is no limit in size of parameters num_consumers, num_queues.

Also, as a local (unique for each table) consistent-hash-exchange is used for sharding within one table and as there still remains a rule that hash-exchange distributes messages based on a hash value of an integer-key, which in our case can be a string, a complicated pattern or a key=value header, then in current implementation hash-property of the exchange is changed to message_id. (This will make it possible to enable sharding with non-integer routing keys). But remember, that to be able to use a different hash-property, by RabbitMQ documentation, when messages are published from some RabbitMQ client, then the message_id property must be set in publishing parameters (otherwise all messages are routed to one arbitrary chosen queue). And to enable sharding - this message_id should be unique for each message or at least unique for every batch of messages.

(!) If there is no need for a specific exchange type, then there is a default implementation. It should be used, if there is just a need for specific table to receive messages quickly. Because in this case only direct-exchange type is used, which is the quickest type of all, and for sharding consistent-hash exchange is not used - everything is done only by direct-exchange. This makes a certain difference in speed as direct-exchange works quicker than consistent-hash-exchange. Sharding with direct-exchange is possible because of the special internal bindings, which are made by default. Therefore, with default implementation, a much faster sharding between all concurrent consumers (~ their queues) is done within each table. This implementation is also used in INSERT query.

To use this default behaviour, exchange_type parameter must not be used, exchange_name must be different for each table, there is no need to set routing_key_list parameter (as it will be ignored). Also when publishing messages - the routing key message parameter must be a string integer, randomized in range [1, num_consumers] for every batch of messages (the smaller the batch size, the better). And if num_queues is set, then in range [1, num_consumers * num_queues]. (In this case, the exchange type, where messages are first published by the client is fanout-exchange, later it will be routed to the needed local table's exchange.)

If transactional_channel parameter is set to 1 (true), then publishing inside INSERT queries implementation will be wrapped in transactions.

Note:
- Consistent-hash exchange must be enabled with rabbitmq plugin.
- By RabbitMQ documentation, if there are multiple consumers, then the order of messages is not guaranteed.

### CODE REVIEW

File: src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        , num_queues(num_queues_)
+        , stopped(stopped_)
+{
+    /* It turned out to be very important to make a different connection each time the object of this class is created,
+     * because in case when num_consumers > 1 - inputStreams run asynchronously and if they share the same connection,
+     * then they also will share the same event loop. But it will mean that if one stream's consumer starts event loop,
+     * then it will run all callbacks on the connection - including other stream's consumer's callbacks - 
+     * it result in asynchronous run of the same code and lead to occasional seg faults.
+     */
+    while (!connection.ready())
```
Comment (alesapin) [inline src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp:43]: infinite loop, need some counter and sleeps.

File: src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        char row_delimiter_,
+        const bool bind_by_id_,
+        const bool hash_exchange_,
+        const size_t num_queues_,
+        const std::atomic<bool> & stopped_)
+        : ReadBuffer(nullptr, 0)
+        , evbase(event_base_new())
+        , eventHandler(evbase, log)
+        , connection(&eventHandler, 
+          AMQP::Address(parsed_address.first, parsed_address.second, AMQP::Login("root", "clickhouse"), "/"))
```
Comment (alesapin) [inline src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp:26]: I think we should remove hardcoded login/password :)

File: src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+            LOG_ERROR(log, "Failed to create queue binding: " << message);
+        });
+    })
+    .onError([&](const char * message)
+    {
+        bindings_error = true;
+        LOG_ERROR(log, "Failed to declare queue on the channel: " << message);
+    });
+
+    while (!bindings_ok && !bindings_error)
```
Comment (alesapin) [inline src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp:162]: These variables can be updated in a separate thread? Need a comment about, now it seems like endless loop :)

File: src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        if (received.empty())
+        {
+            /* Run the onReceived callbacks to save the messages that have been received by now
+             */
+            startNonBlockEventLoop();
+        }
+
+        if (received.empty())
+        {
+            LOG_TRACE(log, "Stalled");
```
Comment (alesapin) [inline src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp:250]: bad message.

File: src/Storages/RabbitMQ/StorageRabbitMQ.cpp [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        , parsed_address(parseAddress(global_context.getMacros()->expand(host_port_), 5672))
+{
+    rabbitmq_context.makeQueryContext();
+
+    setColumns(columns_);
+    task = global_context.getSchedulePool().createTask(log->name(), [this]{ threadFunc(); });
+    task->deactivate();
+
+    /// Enable a different routing algorithm.
+    bind_by_id = num_consumers > 1 || num_queues > 1 || bind_by_id;
```
Comment (alesapin) [inline src/Storages/RabbitMQ/StorageRabbitMQ.cpp:83]: Already initialized above.

File: src/Storages/RabbitMQ/StorageRabbitMQ.cpp [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    for (size_t i = 0; i < num_consumers; ++i)
+    {
+        try
+        {
+            pushReadBuffer(createReadBuffer());
+            ++num_created_consumers;
+        }
+        catch (const AMQP::Exception &)
+        {
+            tryLogCurrentException(log);
```
Comment (alesapin) [inline src/Storages/RabbitMQ/StorageRabbitMQ.cpp:123]: Why we just write log message about exception here? Maybe we should throw?

File: src/Storages/RabbitMQ/StorageRabbitMQ.cpp [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+
+void StorageRabbitMQ::shutdown()
+{
+    stream_cancelled = true;
+
+    for (size_t i = 0; i < num_created_consumers; ++i)
+    {
+        auto buffer = popReadBuffer();
```
Comment (alesapin) [inline src/Storages/RabbitMQ/StorageRabbitMQ.cpp:137]: unused variable

File: src/Storages/RabbitMQ/StorageRabbitMQ.cpp [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    /// Wait for attached views
+    if (!stream_cancelled)
+        task->scheduleAfter(500);
```
Comment (alesapin) [inline src/Storages/RabbitMQ/StorageRabbitMQ.cpp:247]: What is 500 ms? Why not immediately?

**Review by alesapin — COMMENTED** [review]
Almost ready! Now we need to fix style check https://clickhouse-test-reports.s3.yandex.net/11069/c3569882bbcc33c047d9d9b9424bf06b9a50a3bf/style_check.html#fail1 and clang-tidy warnings https://clickhouse-builds.s3.yandex.net/11069/c3569882bbcc33c047d9d9b9424bf06b9a50a3bf/build_log_680227519_1590004367.txt. After that I'll try to help with compatibility. 

Do you have any ideas why some kafka tests failed in your PR?

File: src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+}
+
+
+void ReadBufferFromRabbitMQConsumer::startEventLoop(std::atomic<bool> & loop_started)
+{
+    event_handler.startConsumerLoop(loop_started);
+}
+
+
+bool ReadBufferFromRabbitMQConsumer::nextImpl()
```
Comment (alesapin) [inline src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp:434]: This logic is very complicated and error-prone. We discussed in chat, that it's better to create a separate task to the which will run non-blocking loop iterations and store records in `ConcurrentBoundedQueue` until it's full. Reader threads will just pop messages from the queue and store them in a local buffer. The code should become much simpler.

File: src/Storages/RabbitMQ/RabbitMQBlockInputStream.cpp [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+        auto exchange_name = buffer->getExchange();
+
+        for (size_t i = 0; i < new_rows; ++i)
+        {
+            virtual_columns[0]->insert(exchange_name);
+        }
+
+        total_rows = total_rows + new_rows;
+        buffer->allowNext();
```
Comment (alesapin) [inline src/Storages/RabbitMQ/RabbitMQBlockInputStream.cpp:138]: Suspicious. How buffer will work if we exit the loop in `->eof` branch? As I understand we reuse them.

File: src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+        {
+            /// Run the onReceived callbacks to save the messages that have been received by now, blocks current thread.
+            startEventLoop(loop_started);
+            loop_started.store(false);
+        }
+
+        if (received.empty())
+            return false;
+
+        messages.clear();
```
Comment (alesapin) [inline src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp:451]: race condition with line below (we should work with buffers under locks?)

File: src/Storages/RabbitMQ/StorageRabbitMQ.cpp [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+    uv_loop_init(loop.get());
+
+    event_handler = std::make_unique<RabbitMQHandler>(loop.get(), log);
+    connection = std::make_unique<AMQP::TcpConnection>(event_handler.get(), AMQP::Address(parsed_address.first, parsed_address.second, AMQP::Login(login_password.first, login_password.second), "/"));
+
+    size_t cnt_retries = 0;
+    while (!connection->ready() && ++cnt_retries != Connection_setup_retries_max)
+    {
+        uv_run(loop.get(), UV_RUN_NOWAIT);
+        std::this_thread::sleep_for(std::chrono::milliseconds(Connection_setup_sleep));
```
Comment (alesapin) [inline src/Storages/RabbitMQ/StorageRabbitMQ.cpp:96]: Race condition in AMQCPP triggered here: (write)[https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/src/linux_tcp/tcpresolver.h#L100] , (read)[https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/src/linux_tcp/tcpresolver.h#L200]. Strange async code  https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/src/linux_tcp/tcpresolver.h#L166.

**Review by alesapin — COMMENTED** [review]
We discussed that we have to simplify the code of `(Read/Write)Buffer` with separation of the logic of reading from RabbitMQ (and moving it into separate background task) and reading into views/select queries.

**Review by alesapin — APPROVED** [review]
We still have several points to improve and clarify. However this code is tested well and quite isolated, so I think we can merge it and perform some improvements in separate PRs.

**Comment by qza1800** [comment]
@kssenii this feature is really fantastic
Can you please give a sample how to use `Protobuf` format? I can't find a way to put `format_schema`
https://clickhouse.tech/docs/en/interfaces/formats/#protobuf
Thank you

**Comment by kssenii** [comment]
@qza1800 thank you. Now `format_schema` is not added. I will add it here https://github.com/ClickHouse/ClickHouse/pull/12761.

**Pull request merged**